In [1]:
import pandas as pd
import numpy as np
import os
import re

In [2]:
os.chdir('/Users/chrissoria/Documents/Research/UCNets_Classification/Hand_Coding_Surveys/a19i')
current_directory = os.getcwd()
print(current_directory)

/Users/chrissoria/Documents/Research/UCNets_Classification/Hand_Coding_Surveys/a19i


In [3]:
Claude = pd.read_excel("a19i_Master.xlsx", engine='openpyxl', sheet_name="Claude").drop(["Comments", "Response","prim_key"], axis=1)
Chris = pd.read_excel("a19i_Master.xlsx", engine='openpyxl', sheet_name="Chris").drop(["Comments", "Response"], axis=1)
Rebeca = pd.read_excel("a19i_Master.xlsx", engine='openpyxl', sheet_name="Rebeca").drop(["Comments", "Response"], axis=1)

In [4]:
Claude = Claude.fillna(0)
Claude = Claude.applymap(lambda x: int(x) if isinstance(x, (int, float)) else x)
Chris = Chris.fillna(0)
Chris = Chris.applymap(lambda x: int(x) if isinstance(x, (int, float)) else x)
Rebeca = Rebeca.fillna(0)
Rebeca = Rebeca.applymap(lambda x: int(x) if isinstance(x, (int, float)) else x)


Claude.head()

Partner or spouse  Relationship end  Own job/school/career  \
0                  0                 0                      0   
1                  1                 0                      0   
2                  1                 0                      0   
3                  0                 0                      1   
4                  0                 0                      0   

   Partner’s job/school/career  Financial  Housing, concretely  \
0                            0          0                    0   
1                            0          0                    0   
2                            0          0                    0   
3                            0          0                    0   
4                            0          0                    1   

   New living environment  Family Support  Meet and Make Friends  Conflicts  \
0                       0               1                      0          0   
1                       0               0                      0          0   
2                       0               0                      0          0   
3                       1               0                      0          0   
4                       0               1                      0          0   

   Forced to move  Other reason  Not clear  
0               0             0          0  
1               0             0          0  
2               0             0          0  
3               0             0          0  
4               0             0          0

In [5]:
Claude['cf_category_id'] = Claude.apply(lambda row: ''.join(row.astype(str)), axis=1)
Chris['cs_category_id'] = Chris.apply(lambda row: ''.join(row.astype(str)), axis=1)
Rebeca['re_category_id'] = Rebeca.apply(lambda row: ''.join(row.astype(str)), axis=1)
Claude

Partner or spouse  Relationship end  Own job/school/career  \
0                    0                 0                      0   
1                    1                 0                      0   
2                    1                 0                      0   
3                    0                 0                      1   
4                    0                 0                      0   
..                 ...               ...                    ...   
418                  0                 0                      0   
419                  0                 0                      0   
420                  0                 0                      0   
421                  1                 0                      0   
422                  0                 0                      0   

     Partner’s job/school/career  Financial  Housing, concretely  \
0                              0          0                    0   
1                              0          0                    0   
2                              0          0                    0   
3                              0          0                    0   
4                              0          0                    1   
..                           ...        ...                  ...   
418                            0          0                    0   
419                            0          0                    0   
420                            0          1                    0   
421                            0          0                    0   
422                            0          0                    1   

     New living environment  Family Support  Meet and Make Friends  Conflicts  \
0                         0               1                      0          0   
1                         0               0                      0          0   
2                         0               0                      0          0   
3                         1               0                      0          0   
4                         0               1                      0          0   
..                      ...             ...                    ...        ...   
418                       0               0                      0          0   
419                       0               1                      0          0   
420                       0               0                      0          0   
421                       0               0                      0          0   
422                       0               0                      0          0   

     Forced to move  Other reason  Not clear cf_category_id  
0                 0             0          0  0000000100000  
1                 0             0          0  1000000000000  
2                 0             0          0  1000000000000  
3                 0             0          0  0010001000000  
4                 0             0          0  0000010100000  
..              ...           ...        ...            ...  
418               0             1          0  0000000000010  
419               0             0          0  0000000100000  
420               0             0          0  0000100000000  
421               0             0          0  1000000000000  
422               0             0          0  0000010000000  

[423 rows x 14 columns]

In [6]:
Claude = pd.merge(Claude, Chris[['cs_category_id']], left_index=True, right_index=True)
Claude = pd.merge(Claude, Rebeca[['re_category_id']], left_index=True, right_index=True)

Below, I'm generating measures of agreement, where if two out of three people agree on a certain coding scheme, we will use it. I'm also generating a varible for identifying where we all disagree where I will go through manually and check which of the three works best.

In [7]:
Claude['All_Match'] = Claude.apply(lambda row: 0 if row['cs_category_id'] != row['cf_category_id'] or row['cs_category_id'] != row['re_category_id'] or row['cf_category_id'] != row['re_category_id'] else 1, axis=1)
Claude['CF_RE_Match'] = Claude.apply(lambda row: 1 if row['re_category_id'] == row['cf_category_id'] else 0, axis = 1)
Claude['CF_CS_Match'] = Claude.apply(lambda row: 1 if row['cf_category_id'] == row['cs_category_id'] else 0, axis = 1)
Claude['RE_CS_Match'] = Claude.apply(lambda row: 1 if row['re_category_id'] == row['cs_category_id'] else 0, axis = 1)
Claude['No_Match'] = Claude.apply(lambda row: 1 if row['All_Match'] == 0 and row['CF_RE_Match'] == 0 and row['CF_CS_Match'] == 0 and row['RE_CS_Match'] == 0 else 0, axis=1)

Claude

Partner or spouse  Relationship end  Own job/school/career  \
0                    0                 0                      0   
1                    1                 0                      0   
2                    1                 0                      0   
3                    0                 0                      1   
4                    0                 0                      0   
..                 ...               ...                    ...   
418                  0                 0                      0   
419                  0                 0                      0   
420                  0                 0                      0   
421                  1                 0                      0   
422                  0                 0                      0   

     Partner’s job/school/career  Financial  Housing, concretely  \
0                              0          0                    0   
1                              0          0                    0   
2                              0          0                    0   
3                              0          0                    0   
4                              0          0                    1   
..                           ...        ...                  ...   
418                            0          0                    0   
419                            0          0                    0   
420                            0          1                    0   
421                            0          0                    0   
422                            0          0                    1   

     New living environment  Family Support  Meet and Make Friends  Conflicts  \
0                         0               1                      0          0   
1                         0               0                      0          0   
2                         0               0                      0          0   
3                         1               0                      0          0   
4                         0               1                      0          0   
..                      ...             ...                    ...        ...   
418                       0               0                      0          0   
419                       0               1                      0          0   
420                       0               0                      0          0   
421                       0               0                      0          0   
422                       0               0                      0          0   

     ...  Other reason  Not clear  cf_category_id cs_category_id  \
0    ...             0          0   0000000100000  0000000100000   
1    ...             0          0   1000000000000  1000000000000   
2    ...             0          0   1000000000000  1000000000000   
3    ...             0          0   0010001000000  0010001000000   
4    ...             0          0   0000010100000  0000000000010   
..   ...           ...        ...             ...            ...   
418  ...             1          0   0000000000010  0000000000000   
419  ...             0          0   0000000100000  0000000100000   
420  ...             0          0   0000100000000  0000100000000   
421  ...             0          0   1000000000000  1000000000000   
422  ...             0          0   0000010000000  0000010000010   

    re_category_id All_Match  CF_RE_Match  CF_CS_Match  RE_CS_Match  No_Match  
0    0000000100000         1            1            1            1         0  
1    1000000000000         1            1            1            1         0  
2    1000000000000         1            1            1            1         0  
3    0010001000000         1            1            1            1         0  
4    0000001000000         0            0            0            0         1  
..             ...       ...          ...          ...          ...       ...  
418  0000000000010         0     

The protocal for deciding on a categorization will be if all three agree, then use Claude. \
If Claude and Chris agree, use Claude. \
If Claude and Rebeca agree, use Claude. \
If Chris and Rebeca agree, use Chris. \
If none match, use Chris (and then I will manually review these myself. \
There were only 44 places where we all three disagreed. \
After GPT there were 29.

In [8]:
columns_to_replace = Claude.columns[0:12]

Claude.loc[Claude['RE_CS_Match'] == 1, columns_to_replace] = Chris.loc[Claude.index[Claude['RE_CS_Match'] == 1], columns_to_replace]
Claude.loc[Claude['No_Match'] == 1, columns_to_replace] = Chris.loc[Claude.index[Claude['No_Match'] == 1], columns_to_replace]

In [9]:
Response = pd.read_excel("a19i_Master.xlsx", engine='openpyxl', sheet_name="Claude")
Claude['Response'] = Response['Response']

In [ ]:
# Read the response column from the Excel file
response = pd.read_excel("a19i_Master.xlsx", engine='openpyxl', sheet_name="Chris", usecols="A")

# Rename the column to "Response"
response = response.rename(columns={"A": "Response"})

# Add the response column as the first column in the 'Chris' DataFrame
Chris.insert(0, "Response", response)

Chris.to_excel("Chris_with_correlations3.xlsx", index=False)
Chris

In [10]:
Claude.to_csv('Master_Codes_a19i.csv', index = False)

In [26]:
import openai
from langchain.chat_models import ChatOpenAI
from langchain.prompts import ChatPromptTemplate
from langchain.prompts.prompt import PromptTemplate #where we change the AI "personality"
from dotenv import load_dotenv, find_dotenv
import json

In [12]:
os.chdir('/Users/chrissoria/Documents/Research/Categorization_AI_experiments')

load_dotenv()
API_KEY = os.environ.get("OPENAI_API_KEY")

_ = load_dotenv(find_dotenv()) # read local .env file

#openai.api_key = os.getenv("api.txt")
COMPLETIONS_MODEL = "text-davinci-002"
BETTER_COMPLETIONS_MODEL = "text-davinci-003" #for my purposes, this is better
LONG_MODEL = "gpt-3.5-turbo-16k"
REGULAR_MODEL = "gpt-3.5-turbo"
GPT_4 = "gpt-4"

chat = ChatOpenAI(temperature=0.0,
                  openai_api_key = API_KEY,
                  verbose=True,
                  model_name=REGULAR_MODEL) #depending on how big of a task

#below, we give the AI a "personality"
template = """The following is a conversation between a human data scientist and an AI who specializes in data categorization. The AI is direct and provides concise responses. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
{history}
Data Scientist: {input}
AI:"""

chat

ChatOpenAI(cache=None, verbose=True, callbacks=None, callback_manager=None, tags=None, metadata=None, client=<class 'openai.api_resources.chat_completion.ChatCompletion'>, model_name='gpt-3.5-turbo', temperature=0.7, model_kwargs={}, openai_api_key='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', openai_api_base='', openai_organization='', openai_proxy='', request_timeout=None, max_retries=6, streaming=False, n=1, max_tokens=None, tiktoken_model_name=None)

In [14]:
No_Match = Claude[Claude['No_Match'] == 1]

In [18]:
No_Match

Partner or spouse  Relationship end  Own job/school/career  \
4                    0                 0                      0   
25                   0                 0                      0   
33                   0                 0                      0   
38                   0                 0                      0   
52                   0                 0                      0   
65                   0                 0                      0   
71                   0                 0                      0   
76                   0                 0                      1   
96                   0                 0                      0   
100                  0                 0                      1   
109                  0                 0                      0   
110                  0                 0                      0   
124                  0                 0                      1   
130                  0                 0                      0   
138                  0                 0                      0   
143                  0                 1                      0   
150                  0                 0                      0   
153                  0                 0                      0   
178                  0                 0                      0   
188                  0                 0                      0   
189                  0                 0                      0   
215                  0                 0                      0   
222                  0                 0                      0   
223                  0                 0                      0   
228                  0                 0                      1   
237                  0                 0                      0   
239                  0                 0                      0   
246                  0                 1                      1   
279                  0                 0                      0   
285                  0                 0                      0   
295                  0                 1                      0   
328                  0                 0                      0   
346                  0                 0                      0   
356                  0                 0                      0   
360                  0                 0                      0   
375                  0                 0                      1   
376                  1                 0                      1   
379                  0                 0                      0   
385                  0                 0                      0   
397                  0                 0                      0   
400                  0                 0                      1   
410                  0                 0                      0   
415                  0                 0                      0   

     Partner’s job/school/career  Financial  Housing, concretely  \
4                              0          0                    0   
25                             0          0                    0   
33                             0          0                    1   
38                             0          0                    0   
52                             0          0                    0   
65                             0          0                    0   
71                             0          1                    1   
76                             0          0                    0   
96                             0          0                    0   
100                            0          1                    0   
109                            0          1                    0   
110                            0          0                    0   
124                            0          0                    0   
130                            0          1                    0   
138                            0    

In [21]:
survey_input = No_Match['Response'][25] 

category = """1. to start living with or to stay with partner/spouse. \
2. relationship change (divorce, breakup, etc). \
3. the respondent had a job/school/career change, including transferred. \
4. their partner had a job/school/career change, including transferred. \
5. financial reasons (rent is too expensive, pay raise, etc).
6. specific housing features and preferences, different housing conditions \
7. the desire for a new living environment or neighborhood. \
8. moving closer to family members for various purposes. \
9. being closer to friends or expanding one's social circle. \
10. conflicts with living mates, family members, landlord, or neighbors. \
11. compelled to move due to circumstances beyond their control, such as being evicted. \
12. any other reason for moving that does not fit into any predefined category above. \
13. a responses that does not provide sufficient content or meaningful information to determine a valid reason for moving."""

example_JSON = """{ \
"1": "0", \
"2": "1", \
"3": "0", \
"4": 1, \
"5": "0", \
"6": "0",
"7": "0",
"8": "1",
"9": "0",
"10": "0",
"11": "0",
"12": "0",
"13": "0"
}"""

template_string1 = """A survey respondent was asked, "Why did you move?" \
They responded with: "{OBJECT}" \
First, filter out anything in this response that doesn't answer the question, "Why did you move?" \
Second, pull out all the major reasons why they moved without being too specific. \
Format your response in a as few words as possible starting with the words, 'This respondent moved because...'"""

prompt_template1 = ChatPromptTemplate.from_template(template_string1)
prompt_template1.messages[0].prompt #this will show us our prompt template

GPT_Responses1 = prompt_template1.format_messages(
                    OBJECT=survey_input)

TEST1 = chat(GPT_Responses1)
TEST1 = TEST1.content

template_string2 = """A survey respondent was asked, "Why did you move?" \
"{OBJECT}" \
Please determine how many of the following reasons for moving they provide from this list: \
{CATEGORY} \
Next, provide your answer as a 1 if yes and a 0 if no in JSON format \
Here's an example of how you should format your response: \
{EXAMPLE}"""

prompt_template2 = ChatPromptTemplate.from_template(template_string2)
prompt_template2.messages[0].prompt #this will show us our prompt template

GPT_Responses2 = prompt_template2.format_messages(
                    OBJECT=TEST1,
                    CATEGORY=category,
                    EXAMPLE=example_JSON)

TEST2 = chat(GPT_Responses2)
TEST2 = TEST2.content
print(TEST1)
print(TEST2)

This respondent moved because they didn't like the place and found it difficult due to strange new people and lots of rules.
{ "1": "0", "2": "0", "3": "0", "4": "0", "5": "0", "6": "0", "7": "1", "8": "0", "9": "0", "10": "1", "11": "0", "12": "0", "13": "0" }


In [27]:
for index, row in No_Match.iterrows():
    survey_input = row[survey_participant_input]
    response = prompt_template1.format_messages(
                    OBJECT=survey_input)

    response = chat(response)
    
    # Assuming a successful attempt means a non-empty response
    if response.content:
        print(f"Successful attempt for row number for chain 1: {index}")
    
    No_Match.at[index, 'Key_Reasons'] = response.content

for index, row in No_Match.iterrows():
    survey_input = row['Key_Reasons']
    response = prompt_template2.format_messages(
                    OBJECT=survey_input,
                    CATEGORY=category,
                    EXAMPLE=example_JSON)

    response = chat(response)
    
    # Assuming a successful attempt means a non-empty response
    if response.content:
        print(f"Successful attempt for row number for chain 2: {index}")
    
    No_Match.at[index, 'JSON'] = response.content

Successful attempt for row number for chain 1: 4
Successful attempt for row number for chain 1: 25
Successful attempt for row number for chain 1: 33
Successful attempt for row number for chain 1: 38
Successful attempt for row number for chain 1: 52
Successful attempt for row number for chain 1: 65
Successful attempt for row number for chain 1: 71
Successful attempt for row number for chain 1: 76
Successful attempt for row number for chain 1: 96
Successful attempt for row number for chain 1: 100
Successful attempt for row number for chain 1: 109
Successful attempt for row number for chain 1: 110
Successful attempt for row number for chain 1: 124
Successful attempt for row number for chain 1: 130
Successful attempt for row number for chain 1: 138
Successful attempt for row number for chain 1: 143
Successful attempt for row number for chain 1: 150
Successful attempt for row number for chain 1: 153
Successful attempt for row number for chain 1: 178
Successful attempt for row number for cha

In [33]:
normalized_data_list = []
error_lines = []

for i, json_str in enumerate(No_Match['JSON']):
    try:
        parsed_obj = json.loads(json_str)
        normalized_data_list.append(pd.json_normalize(parsed_obj))
    except json.JSONDecodeError:
        error_lines.append(i)
        continue

# Concatenate the normalized data into one DataFrame
normalized_data = pd.concat(normalized_data_list, ignore_index=True)
normalized_data.index = No_Match.index

error_lines

[]

In [34]:
normalized_data

1  2  3  4  5  6  7  8  9 10 11 12 13
4    0  0  0  0  0  0  0  0  0  0  0  0  0
25   0  0  0  0  0  0  0  0  0  1  0  0  0
33   0  0  0  0  0  1  0  0  0  0  0  0  0
38   0  0  0  0  0  1  0  0  0  1  0  0  0
52   0  1  0  0  0  0  0  0  0  1  0  0  0
65   1  0  0  0  0  0  0  0  0  0  0  0  0
71   0  0  0  0  0  0  0  0  0  0  0  1  0
76   0  0  1  0  0  0  1  0  0  0  0  0  0
96   0  0  0  0  0  0  0  0  0  0  0  0  0
100  0  0  0  0  0  0  0  1  1  0  0  0  0
109  0  0  0  0  1  0  0  0  0  0  0  0  0
110  0  0  0  0  0  0  1  0  0  0  0  0  0
124  0  0  1  0  0  0  1  0  0  0  0  0  0
130  0  0  0  0  0  1  0  0  0  0  0  0  0
138  1  0  0  0  0  0  0  0  0  0  0  0  0
143  0  0  0  0  0  0  0  0  1  0  0  0  0
150  0  0  0  0  0  0  0  0  0  0  0  0  0
153  0  0  0  0  0  0  0  0  0  0  0  0  0
178  0  0  0  0  0  0  0  0  0  0  0  0  0
188  0  0  0  0  1  0  0  0  0  0  0  0  0
189  0  0  0  0  0  0  0  0  0  0  0  0  0
215  0  0  0  0  1  1  0  0  0  0  0  0  0
222  0  0  0  0  0  0  0  0  0  0  0  0  0
223  0  0  0  0  0  0  0  0  0  0  1  0  0
228  0  0  1  0  0  0  0  0  0  0  0  0  0
237  0  0  0  0  0  0  0  0  1  0  0  0  0
239  0  0  0  0  0  0  0  0  0  1  0  0  0
246  0  1  1  0  0  0  0  0  0  0  0  0  0
279  0  0  0  0  0  0  0  0  1  0  0  0  0
285  0  0  0  0  0  0  0  0  0  0  0  0  0
295  0  1  0  0  1  0  0  0  0  0  0  0  0
328  0  0  0  0  0  1  0  0  0  0  0  0  0
346  0  0  0  0  0  0  0  1  0  0  0  0  0
356  0  0  0  0  0  0  0  0  0  1  0  0  0
360  0  0  0  1  0  0  0  0  0  0  0  0  0
375  0  0  1  0  1  0  0  0  0  0  0  0  0
376  0  0  0  0  0  1  0  0  0  0  0  0  0
379  1  0  0  0  0  0  0  0  0  0  0  0  0
385  0  0  0  0  1  0  0  0  0  0  0  0  0
397  0  0  0  0  0  0  0  0  0  0  0  0  0
400  0  0  1  0  0  0  1  0  0  0  0  0  0
410  0  0  0  0  0  1  0  0  0  1  0  0  0
415  0  0  0  0  0  0  0  1  0  1  0  0  0

In [35]:
No_Match['gpt_category_id'] = normalized_data.apply(lambda row: ''.join(row.astype(str)), axis=1)
No_Match

/var/folders/89/6bdxzk2j30v5n3wstywbcpg80000gn/T/ipykernel_88215/2444947372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  No_Match['gpt_category_id'] = normalized_data.apply(lambda row: ''.join(row.astype(str)), axis=1)


Partner or spouse  Relationship end  Own job/school/career  \
4                    0                 0                      0   
25                   0                 0                      0   
33                   0                 0                      0   
38                   0                 0                      0   
52                   0                 0                      0   
65                   0                 0                      0   
71                   0                 0                      0   
76                   0                 0                      1   
96                   0                 0                      0   
100                  0                 0                      1   
109                  0                 0                      0   
110                  0                 0                      0   
124                  0                 0                      1   
130                  0                 0                      0   
138                  0                 0                      0   
143                  0                 1                      0   
150                  0                 0                      0   
153                  0                 0                      0   
178                  0                 0                      0   
188                  0                 0                      0   
189                  0                 0                      0   
215                  0                 0                      0   
222                  0                 0                      0   
223                  0                 0                      0   
228                  0                 0                      1   
237                  0                 0                      0   
239                  0                 0                      0   
246                  0                 1                      1   
279                  0                 0                      0   
285                  0                 0                      0   
295                  0                 1                      0   
328                  0                 0                      0   
346                  0                 0                      0   
356                  0                 0                      0   
360                  0                 0                      0   
375                  0                 0                      1   
376                  1                 0                      1   
379                  0                 0                      0   
385                  0                 0                      0   
397                  0                 0                      0   
400                  0                 0                      1   
410                  0                 0                      0   
415                  0                 0                      0   

     Partner’s job/school/career  Financial  Housing, concretely  \
4                              0          0                    0   
25                             0          0                    0   
33                             0          0                    1   
38                             0          0                    0   
52                             0          0                    0   
65                             0          0                    0   
71                             0          1                    1   
76                             0          0                    0   
96                             0          0                    0   
100                            0          1                    0   
109                            0          1                    0   
110                            0          0                    0   
124                            0          0                    0   
130                            0          1                    0   
138                            0    

In [36]:
gpt_category_id = No_Match['gpt_category_id']
gpt_category_id

4      0000000000000
25     0000000001000
33     0000010000000
38     0000010001000
52     0100000001000
65     1000000000000
71     0000000000010
76     0010001000000
96     0000000000000
100    0000000110000
109    0000100000000
110    0000001000000
124    0010001000000
130    0000010000000
138    1000000000000
143    0000000010000
150    0000000000000
153    0000000000000
178    0000000000000
188    0000100000000
189    0000000000000
215    0000110000000
222    0000000000000
223    0000000000100
228    0010000000000
237    0000000010000
239    0000000001000
246    0110000000000
279    0000000010000
285    0000000000000
295    0100100000000
328    0000010000000
346    0000000100000
356    0000000001000
360    0001000000000
375    0010100000000
376    0000010000000
379    1000000000000
385    0000100000000
397    0000000000000
400    0010001000000
410    0000010001000
415    0000000101000
Name: gpt_category_id, dtype: object

In [37]:
Claude_GPT = pd.concat([Claude, gpt_category_id], axis=1)
Claude_GPT

Partner or spouse  Relationship end  Own job/school/career  \
0                    0                 0                      0   
1                    1                 0                      0   
2                    1                 0                      0   
3                    0                 0                      1   
4                    0                 0                      0   
..                 ...               ...                    ...   
418                  0                 0                      0   
419                  0                 0                      0   
420                  0                 0                      0   
421                  1                 0                      0   
422                  0                 0                      0   

     Partner’s job/school/career  Financial  Housing, concretely  \
0                              0          0                    0   
1                              0          0                    0   
2                              0          0                    0   
3                              0          0                    0   
4                              0          0                    0   
..                           ...        ...                  ...   
418                            0          0                    0   
419                            0          0                    0   
420                            0          1                    0   
421                            0          0                    0   
422                            0          0                    1   

     New living environment  Family Support  Meet and Make Friends  Conflicts  \
0                         0               1                      0          0   
1                         0               0                      0          0   
2                         0               0                      0          0   
3                         1               0                      0          0   
4                         0               0                      0          0   
..                      ...             ...                    ...        ...   
418                       0               0                      0          0   
419                       0               1                      0          0   
420                       0               0                      0          0   
421                       0               0                      0          0   
422                       0               0                      0          0   

     ...  cf_category_id  cs_category_id  re_category_id All_Match  \
0    ...   0000000100000   0000000100000   0000000100000         1   
1    ...   1000000000000   1000000000000   1000000000000         1   
2    ...   1000000000000   1000000000000   1000000000000         1   
3    ...   0010001000000   0010001000000   0010001000000         1   
4    ...   0000010100000   0000000000010   0000001000000         0   
..   ...             ...             ...             ...       ...   
418  ...   0000000000010   0000000000000   0000000000010         0   
419  ...   0000000100000   0000000100000   0000000100000         1   
420  ...   0000100000000   0000100000000   0000100000000         1   
421  ...   1000000000000   1000000000000   1000000000000         1   
422  ...   0000010000000   0000010000010   0000010000000         0   

    CF_RE_Match CF_CS_Match  RE_CS_Match  No_Match  \
0             1           1            1         0   
1             1           1            1         0   
2             1           1            1         0   
3             1           1            1         0   
4             0           0            0         1   
..          ...         ...          ...       ...   
418           1           0            0         0   
419           1           1            1         0   
420           1           1            1         0   
421           1           1  

In [38]:
Claude_GPT.to_csv('/Users/chrissoria/Documents/Research/UCNets_Classification/Hand_Coding_Surveys/a19i/Claude_GPT.csv')

In [42]:
Claude_GPT['CF_GP_Match'] = Claude_GPT.apply(lambda row: 1 if row['cf_category_id'] == row['gpt_category_id'] else 0, axis = 1)
Claude_GPT['CS_GP_Match'] = Claude_GPT.apply(lambda row: 1 if row['cs_category_id'] == row['gpt_category_id'] else 0, axis = 1)
Claude_GPT['RE_GP_Match'] = Claude_GPT.apply(lambda row: 1 if row['re_category_id'] == row['gpt_category_id'] else 0, axis = 1)

In [43]:
Claude_GPT

Partner or spouse  Relationship end  Own job/school/career  \
0                    0                 0                      0   
1                    1                 0                      0   
2                    1                 0                      0   
3                    0                 0                      1   
4                    0                 0                      0   
..                 ...               ...                    ...   
418                  0                 0                      0   
419                  0                 0                      0   
420                  0                 0                      0   
421                  1                 0                      0   
422                  0                 0                      0   

     Partner’s job/school/career  Financial  Housing, concretely  \
0                              0          0                    0   
1                              0          0                    0   
2                              0          0                    0   
3                              0          0                    0   
4                              0          0                    0   
..                           ...        ...                  ...   
418                            0          0                    0   
419                            0          0                    0   
420                            0          1                    0   
421                            0          0                    0   
422                            0          0                    1   

     New living environment  Family Support  Meet and Make Friends  Conflicts  \
0                         0               1                      0          0   
1                         0               0                      0          0   
2                         0               0                      0          0   
3                         1               0                      0          0   
4                         0               0                      0          0   
..                      ...             ...                    ...        ...   
418                       0               0                      0          0   
419                       0               1                      0          0   
420                       0               0                      0          0   
421                       0               0                      0          0   
422                       0               0                      0          0   

     ...  All_Match  CF_RE_Match  CF_CS_Match RE_CS_Match No_Match  \
0    ...          1            1            1           1        0   
1    ...          1            1            1           1        0   
2    ...          1            1            1           1        0   
3    ...          1            1            1           1        0   
4    ...          0            0            0           0        1   
..   ...        ...          ...          ...         ...      ...   
418  ...          0            1            0           0        0   
419  ...          1            1            1           1        0   
420  ...          1            1            1           1        0   
421  ...          1            1            1           1        0   
422  ...          0            1            0           0        0   

                                              Response  gpt_category_id  \
0     relocated back to east coast - closer to my sons              NaN   
1                     move in together with my partner              NaN   
2    out of living with my friends, and into living...              NaN   
3    to take a new job in new york city (both becau...              NaN   
4    wanted to live in my own place outside my pare...    0000000000000   
..                                                 ...              ...   
418  moved in with limi to help take care of

In [44]:
Claude_GPT.to_csv('/Users/chrissoria/Documents/Research/UCNets_Classification/Hand_Coding_Surveys/a19i/Claude_GPT.csv')

In [45]:
columns_to_replace = Claude_GPT.columns[0:12]

Claude_GPT.loc[Claude_GPT['CF_GP_Match'] == 1, columns_to_replace] = Claude_GPT.loc[Claude_GPT.index[Claude_GPT['CF_GP_Match'] == 1], columns_to_replace]
Claude_GPT.loc[Claude_GPT['CF_GP_Match'] == 1, columns_to_replace] = Chris.loc[Claude_GPT.index[Claude_GPT['CF_GP_Match'] == 1], columns_to_replace]
Claude_GPT.loc[Claude_GPT['RE_GP_Match'] == 1, columns_to_replace] = Rebeca.loc[Claude_GPT.index[Claude_GPT['RE_GP_Match'] == 1], columns_to_replace]

Now let's update the no match column based on whether GPT matched any of our responses.

In [55]:
Claude_GPT['No_Match'] = Claude_GPT.apply(lambda row: 0 if row['CF_GP_Match'] == 1 or row['CS_GP_Match'] == 1 or row['RE_GP_Match'] == 1 else row['No_Match'], axis=1)

In [56]:
count_of_ones = sum(Claude_GPT['No_Match'] == 1)
reduction_rate = ((44 - count_of_ones) / 44) * 100
print("GPT was able to reduce rates of non-agreement by " + str(round(reduction_rate, 2)) + "%")

GPT was able to reduce rates of non-agreement by 43.18%


In [57]:
Claude_GPT.to_csv('/Users/chrissoria/Documents/Research/UCNets_Classification/Hand_Coding_Surveys/a19i/Master_Codes_a19i.csv', index = False)

I adjusted line 412 to add "financial"
378 I removed own job and partner's job
377 I added new living environment
358 changed to not clear
297 removed "relationship end"
241 changed it away from "forced to move"
239 changed it to "other reason"
155 I added housing features
145 I removed relationship end
126 I removed meet friends